File: packages/java/src/main/java/fbc/Seguranca/AlvoCriptografico.java
package fbc.Seguranca;

/**
 * Contrato que uma cifra precisa cumprir pra ser testada pela suíte de ataques.
 *
 * Os métodos "obrigatórios" (encrypt/decrypt) bastam pros ataques de alto nível.
 * {@code gerarKeystreamBruto} e {@code checksumBruto} expõem as camadas internas
 * para os ataques de baixo nível.
 */
public interface AlvoCriptografico {
    String encrypt(String texto);

    String encrypt(byte[] texto);

    String decrypt(String token);

    /** Prefixo esperado no início de um token válido (ex: "FBC"). */
    String prefixo();

    /** Decompõe um token (sem prefixo e já base64url-decodificado) nos campos. */
    CamposToken decompor(byte[] tokenDecodificado);

    /** Recompõe um token a partir dos campos (inverso de decompor()). */
    byte[] recompor(CamposToken campos);

    /** Gera N bytes de keystream bruto a partir de key+iv+propósito. */
    byte[] gerarKeystreamBruto(byte[] key, byte[] iv, String proposito, int tamanho);

    /** Chama o checksum() interno diretamente. */
    byte[] checksumBruto(byte[] dados, byte[] key);

    /** Uma chave válida pra usar nos testes. */
    String chaveDeTeste();

    /** Tamanho do IV em bytes. */
    int tamanhoIv();

    String base64urlEncode(byte[] data);

    byte[] base64urlDecode(String data);
}

File: packages/go/internal/seguranca/alvo_criptografico.go
package seguranca

// CamposToken são os campos que compõem um token já decodificado do base64url.
type CamposToken struct {
	Integridade []byte
	Ciphertext  []byte
	IV          []byte
}

// AlvoCriptografico é o contrato que uma cifra precisa cumprir para ser testada
// pela suíte de ataques.
//
// Os métodos obrigatórios (Encrypt/Decrypt) bastam para os ataques de alto nível.
// GerarKeystreamBruto e ChecksumBruto expõem as camadas internas para os ataques
// de baixo nível.
type AlvoCriptografico interface {
	Encrypt(texto []byte) (string, error)

	Decrypt(token string) (string, error)

	// Prefixo esperado no início de um token válido (ex: "FBC").
	Prefixo() string

	// Decompor decompõe um token (sem prefixo e já base64url-decodificado).
	Decompor(tokenDecodificado []byte) CamposToken

	// Recompor recompõe um token a partir dos campos (inverso de Decompor).
	Recompor(campos CamposToken) []byte

	// GerarKeystreamBruto gera N bytes de keystream bruto a partir de key+iv+propósito.
	GerarKeystreamBruto(key, iv []byte, proposito string, tamanho int) []byte

	// ChecksumBruto chama o Checksum interno diretamente.
	ChecksumBruto(dados, key []byte) []byte

	// ChaveDeTeste devolve uma chave válida para usar nos testes.
	ChaveDeTeste() string

	// TamanhoIv devolve o tamanho do IV em bytes.
	TamanhoIv() int

	Base64urlEncode(data []byte) string

	Base64urlDecode(data string) ([]byte, error)
}
